### PR TITLE
perf: Initially trim leading and trailing filtered rows

### PR DESCRIPTION
**Pull request by coastalwhite**
This adds a small step before doing a `filter` and trims away the leading and trailing filtered rows. This cost us practically nothing in the cases where there are no items to trim and saves massively where it can trim items. This is for example in cases where we are filtering by at least one sorted column.

Here is a code snippet to show the effectiveness. Before, it had to copy the memory and go into the filter loop. Now, it can just take the fast path. This also holds for sorted columns which is where I think most of the performance will come from.

```python
import numpy as np
import polars as pl
import os
from timeit import timeit

SIZE = 10_000_000

# Threads always cause a great deal of variability on my laptop
os.environ['POLARS_MAX_THREADS'] = '1'

df = pl.DataFrame([
    pl.Series('f', sum(([i] * (SIZE // 10) for i in range(10)), []), pl.Int32),
] + [
    pl.Series(f'data{i}', np.random.randint(0, 500, SIZE), pl.Int64)
    for i in range(100)
])

t = timeit(lambda: df.filter(pl.col.f == 5), number=10)
print(f"{t}s")
```

```console
Before: 1.1825743649969809s
After: 0.07909882599778939s
```

### CODE REVIEW

**Comment by ritchie46** [comment]
Wow! Great idea. And basically free indeed.

File: crates/polars-compute/src/filter/mod.rs [hunk 1]
```diff
@@ -26,6 +26,20 @@ pub fn filter(array: &dyn Array, mask: &BooleanArray) -> Box<dyn Array> {
 }
 
 pub fn filter_with_bitmap(array: &dyn Array, mask: &Bitmap) -> Box<dyn Array> {
+    // Many filters involve filtering values in a subsection of the array. When we trim the leading
+    // and trailing filtered items, we can close in on those items and not have to perform and
+    // thinking about those. The overhead for when there are no leading or trailing filtered values
+    // is very minimal: only a clone of the mask and the array.
+    //
+    // This also allows dispatching to the fast paths way, way, way more often.
+    let mut mask = mask.clone();
```
Comment (ritchie46) [inline crates/polars-compute/src/filter/mod.rs:35]: Maybe we should do this one line 51? After the full and empty check.

Comment (coastalwhite) [reply crates/polars-compute/src/filter/mod.rs:35]: No the effectiveness stems exactly from doing it before. If we trim to an exact full or empty block, we can take the fast path.

Comment (ritchie46) [reply crates/polars-compute/src/filter/mod.rs:35]: Ah, right. :+1:

**Comment by ankithGraviton** [comment]
Sorry for resurrecting an old, merged PR. Just want to understand if the following was true: The optimization purely effects the frequency of the all 1s mask fast path and doesn't at all effect the frequency of taking the all 0s mask fast path?

Basically (the regex) 0*1+0* also is now always a fast path?

Plus additional perf improvement due to the fact that when trimming occurs it leads to overall less work down-stream

**Comment by coastalwhite** [comment]
Exactly.

**Pull request merged**